File: MORM.CSharp/Classes/Comando.cs
namespace MORM.Reppositorio.Classes
{
    public static class Comando
    {
        //-- tipo database

        public static TipoDatabase TipoDatabase { get; set; } = TipoDatabase.Oracle;
        
        //-- tabela
        
        public static TTabela GetTabela(this Type type)        
        {
            if (Type is CollectionItem)
            {
                var collection = Activator.Create(Type) as CollectionItem;
                return collection.GetTabela();
            {
            else
                return null;
        }
        
        //-- campos

        public static TTabela GetCampos()
        {
            if (Type is CollectionItem)
            {
                var collection = Activator.Create(Type) as CollectionItem;
                return collection.GetCampos();
            {
            else
                return null;
        }

        //-- value

        private static string GetValueStr(this object obj, string attributo)
        {
            var value = obj.GetType().GetProperties()
                .FirstOrDefault(x => x.Name == atributo);
            return (
                value == null ? "null" :
                value.GetType() == typeof(bool) ? "'" + ((bool)value ? "T" : "F") + "'" :
                value.GetType() == typeof(DateTime) ? TipoDatabase.GetValueData((DateTime)value) :
                value.GetType() == typeof(decimal) ? value.ToString().Replace(",", ".") :
                value.GetType() == typeof(double) ? value.ToString().Replace(",", ".") :
                value.GetType() == typeof(int) ? value.ToString() :
                value.GetType() == typeof(string) ? "'" + value.ToString().Replace("'", "''") + "'" : value.ToString());
        }

        //-- string

        public static void AddString(ref string str, string val, string seq, string ini = "")
        {
            str += (!string.IsNullOrWhiteSpace(str) ? sep : ini) + val;
        }

        //-- select

        public static string GetSelect(this Type type, string where = "")
        {
            var tabela = type.GetTabela();
            var campos = type.GetCampos();
            var fieldsAtr = string.Empty;
            var fields = string.Empty;

            foreach (var campo in campos)
            {
                AddString(ref fieldsAtr, campo.Atributo + " as \"" + campo.Atributo + "\"", ", ");
                AddString(ref fields, campo.Campo + " as " + campo.Atributo, ", ");
            }

            return 
                $"select {fieldsAtr} from (select {fields} from {tabela.Nome})" +
                (!string.IsNullOrWhiteSpace(where) ? " where " + where : "") ;
        }

        public static string GetSelect(this TObject obj)
        {
            var campos = obj.GetType().GetCampos();
            var where = string.Empty;

            foreach (var campo in campos)
                if (campo.Tipo == CampoTipo.Key)
                    AddString(ref where, campo.Atributo + " = " + obj.GetValueStr(campo.Atributo), " and ");

            return GetSelect(obj.GetType(), where);
        }

        //-- insert

        public static string GetInsert(this TObject obj)
        {
            var tabela = obj.GetType().GetTabela();
            var campos = obj.GetType().GetCampos();
            var campos = string.Empty;
            var values = string.Empty;

            foreach (var campo in campos)
            {
                AddString(ref campos, campo.Campo, ", ");
                AddString(ref values, obj.GetValueStr(campo.Atributo), ", ");
            }

            return 
                $"insert into {tabela.Nome} ({campos}) values ({values})";
        }

        //-- update

        public static string GetUpdate(this TObject obj)
        {
            var tabela = obj.GetType().GetTabela();
            var campos = obj.GetType().GetCampos();
            var sets = string.Empty;
            var where = string.Empty;

            foreach (var campo in campos)
                if (campo.Tipo == CampoTipo.Key)
                    AddString(ref where, campo.Campo + " = " + obj.GetValueStr(campo.Atributo), " and ");
                else
                    AddString(ref sets, campo.Campo + " = " + obj.GetValueStr(campo.Atributo), ", ");

            return
                $"update {tabela.Nome} set {sets} where {where}";
        }

        //-- delete

        public static string GetDelete(this TObject obj)
        {
            var tabela = obj.GetType().GetTabela();
            var campos = obj.GetType().GetCampos();
            var where = string.Empty;

            foreach (var campo in campos)
                if (campo.Tipo == CampoTipo.Key)
                    AddString(ref where, campo.Campo + " = " + obj.GetValueStr(campo.Atributo), " and ");

            return
                $"delete from {tabela.Nome} where {where}";
        }
    }
}

File: MORM.Php/Classes/comando.php
<?php 

require_once("logger.php");

class Comando
{
    public static function GetTabela($class)
    {
        return $class; // $obj->GetTabela();
    }
    
    public static function GetCampo($obj, $atr)
    {
        return $atr; // $obj->GetCampo($atr);
    }

    //-- string
    
    public static function AddString(&$str, $sub, $sep, $ini = "")
    {
        $str = $str . ($str != "" ? $sep : $ini) . $sub;
    }
    
    //-- value
    
    public static function GetValueStr($value)
    {
        return (isset($value) ? "'" . $value . "'" : "null");
    }
    
    //-- select
    
    public static function GetSelect($class, $where = "")
    {
        $METHOD = "Comando.GetSelect";
        Logger::Instance()->Info($METHOD, "class: " . $class);

        $tabela = self::GetTabela($class);
        $fieldsAtr = "";
        $fields = "";
        
        $obj = new $class;
        foreach ($obj as $name => $value)
        {
            self::AddString($fieldsAtr, $name . ' as "' . $name . '"', ", ");
            self::AddString($fields, self::GetCampo($obj, $name) . " as " . $name, ", ");
        }
        
        $sql = 
            "select " . $fieldsAtr . 
            " from (select " . $fields . " from " . $tabela . ") a " .
            ($where != "" ? " where " . $where : "");

        Logger::Instance()->Info($METHOD, "sql: " . $sql);

        return $sql;
    }

    public static function GetSelectObj($obj)
    {
        $METHOD = "Comando.GetSelectObj";
        Logger::Instance()->Info($METHOD, "class: " . get_class($obj));

        $where = "";
        
        foreach ($obj as $name => $value)
            if ($name == "Codigo")
                self::AddString($where, $name . " = " . self::GetValueStr($value), ", ");

        return self::GetSelect(get_class($obj), $where);
    }
    
    //-- insert / update / delete

    public static function GetInsert($obj)    
    {
        $METHOD = "Comando.GetInsert";
        Logger::Instance()->Info($METHOD, "class: " . get_class($obj));

        $tabela = self::GetTabela(get_class($obj));
        $fields = "";
        $values = "";
        
        foreach ($obj as $name => $value)
        {
            self::AddString($fields, self::GetCampo($obj, $name), ", ");
            self::AddString($values, self::GetValueStr($value), ", ");
        }

        $cmd = 
            "insert into " . $tabela . 
            " (" . $fields . 
            ") values (" . $values . ")";

        Logger::Instance()->Info($METHOD, "cmd: " . $cmd);

        return $cmd;
    }
    
    public static function GetUpdate($obj)    
    {
        $METHOD = "Comando.GetUpdate";
        Logger::Instance()->Info($METHOD, "class: " . get_class($obj));

        $tabela = self::GetTabela(get_class($obj));
        $sets = "";
        $where = "";
        
        foreach ($obj as $name => $value)
            if ($name != "Codigo")
                self::AddString($sets, self::GetCampo($obj, $name) . " = " . self::GetValueStr($value), ", ");
            else
                self::AddString($where, self::GetCampo($obj, $name) . " = " . self::GetValueStr($value), " and ");

        $cmd = 
            "update " . $tabela . 
            " sets " . $sets . 
            " where " . $where;

        Logger::Instance()->Info($METHOD, "cmd: " . $cmd);

        return $cmd;
    }

    public static function GetDelete($obj)    
    {
        $METHOD = "Comando.GetDelete";
        Logger::Instance()->Info($METHOD, "class: " . get_class($obj));

        $tabela = self::GetTabela(get_class($obj));
        $where = "";
        
        foreach ($obj as $name => $value)
            if ($name == "Codigo")
                self::AddString($where, self::GetCampo($obj, $name) . " = " . self::GetValueStr($value), " and ");

        $cmd = 
            "delete from " . $tabela . 
            " where " . $where;

        Logger::Instance()->Info($METHOD, "cmd: " . $cmd);

        return $cmd;
    }    
}
?>